### ISSUE TITLE
How to set the current token into the http interceptor?

### ISSUE THREAD

**Issue by softwareEngineerFTE** [damienbod/angular-auth-oidc-client#1213]
As per your code, if I enable the logger after a specific time I get to see in the console (token: false).

How do I capture this false value in my code base.

Below is the auth-config of my application.

```
oidcConfigService.withConfig({
            stsServer: 'test',
            redirectUrl: `${window.location.origin}/resolve`,
            postLogoutRedirectUri: `${window.location.origin}/unauthorized`,
            clientId: 'auth-poc',
            scope: 'open', // 'openid profile ' + your scopes
            responseType: 'code',
            silentRenew: true,
            triggerAuthorizationResultEvent: true,
            postLoginRoute: '/auth',
            forbiddenRoute: '/forbidden',
            unauthorizedRoute: '/unauthorized',
            useRefreshToken: true,
            historyCleanupOff: true,
            issValidationOff: true,
            customParams: { ['a']:'try'},
silentRenewUrl: `${window.location.origin}/silent-renew.html`
        });
```

Can you please let me know if I'm missing something in the config code or a way to subscribe to the function which you let me know if the session got expired. I do not have a logout button or any button on the screen to refresh the token. I have a canActivate guard to check for the token expiration and I need to use the below code or any other code if you suggest to refresh the token.

As per your example in the silent renew, you have a button to refresh the token.

```
refreshSessionCheckSession() {
    console.log('start refreshSession');
    this.oidcSecurityService.authorize();
  }
```

Thanks in advance and appreciate your help.

**Comment by FabianGosebrink** [comment]
Hey, 

> How do I capture this false value in my code base.

[checkAuth](https://nice-hill-002425310.azurestaticapps.net/docs/documentation/public-api#checkauthurl-string-configid-string)

What are you trying to achieve? You are writing

> How do I capture this false value in my code base.

as well as

> a way to subscribe to the function which you let me know if the session got expired

Further, your `refreshSessionCheckSession` is doing an `authorize`, so a login and not a refresh session.

What are you trying to do?

Thanks

**Comment by softwareEngineerFTE** [comment]
Hi Fabian,

Apologies for not being clear.

I want to update the token by setting request headers for the token end points. 

When I switch on the logger to see the console, I noticed after a specific time the auth-oidc fires an event to the tokenEndpoint to get the new token. When this API is triggered I need to set request headers for the endpoint.

on the logger I see below stuff
1) starting silent renew...
2) BEGIN refresh session Authorize
3) RefreshSession Created. adding my autostate: random number (**i want to change this number with a different code**)
4) found refresh response -- at this time it hits a endpoint for the token. **I want to set headers for this endpoint for a proper response and then set it back in the sessionStorage.**

Can you please tell me how to fix the issue? Below is my auth-config (I'm not using silent-renew.html)

oidcConfigService.withConfig({
            stsServer: 'test',
            redirectUrl: `${window.location.origin}/resolve`,
            postLogoutRedirectUri: `${window.location.origin}/unauthorized`,
            clientId: 'auth-poc',
            scope: 'open', // 'openid profile ' + your scopes
            responseType: 'code',
            silentRenew: true,
            triggerAuthorizationResultEvent: true,
            postLoginRoute: '/auth',
            forbiddenRoute: '/forbidden',
            unauthorizedRoute: '/unauthorized',
            useRefreshToken: true,
            historyCleanupOff: true,
            issValidationOff: true,
            customParams: { ['a']:'try'}
        });

**Comment by FabianGosebrink** [comment]
Hi,

thanks for the answer. Honestly, I still do not know what you are trying to achieve sorry :)

So the lib does the refresh automatically for you if you configured it correctly. No need to store this in the session storage for yourself. This is what the lib is for.

If you want to send the token (which the lib handles, saves and refreshes for you) in a request, set it to the headers using the [http-interceptor](https://nice-hill-002425310.azurestaticapps.net/docs/documentation/using-access-tokens#http-interceptor) from the lib. Is that what you need?

Thanks

**Comment by damienbod** [comment]
Hi @softwareEngineerFTE 

You are implementing something very specific here on version 11. I would recommend updating to version 12 sticking to the standards. You want to customize the flows. Would you not consider using a standard flow and standard IDP? What IDP do you use and why do you require this?

If you need to implement this, then you would need to take the src code and implement this yourself. We provide custom query parameters for each step of the flows in version 12, which is not supported in V11. These could be maybe extended. Also with the comment suggestion from @FabianGosebrink 

Hope this helps

Greetings Damien

**Comment by softwareEngineerFTE** [comment]
Hi Fabian,

What I'm trying to say is I need to get a new token from the rest endpoint which is fired automatically from the auth-oidc lib. 

What your library does is it automatically triggers an event to the rest endpoint to get the new token and set it in the sessionStorage.

If you want to send the token (which the lib handles, saves and refreshes for you) in a request, set it to the headers using the http-interceptor from the lib. Is that what you need? 

What I want to achieve is that when the initial token expire after a specific time. I need to tell my httpInterceptor file that the token is expired. 

If I add a config as logLevel: LogLevel.Debug after a specific time when the token expired and the lib tries to renew the token. In console log from the lib it prints "starting silent renew..."

and before this logger prints something like:-
checking: silentRenewRunning: false id_token: true userData: true
has id_token expired: false, 1655686856> 16596965586
silent renew, periodic check finished!

the above logger loops in till the token is expired.

So when the token expires, library automatically fires the event for the new token, I need to tell the httpInterceptor that the token is expired so that I can update the request headers for the endpoint.

checkAuth() triggers is also getting triggered once i.e; on the page load. 

isAuthencated$ return true every time. 

Is there any other event which can help me determine if token got expired?



@damienbod :- due to some constraint I cannot update the app to angular 12. Thanks for the heads up. Once I upgrade it to 12, I will revisit the above feature to make the necessary changes.

**Comment by FabianGosebrink** [comment]
Hey, 

thanks for the answer. I think there is a misunderstanding in how interceptors work. Interceptors are being called when a request fires. Think of it as a middleware between your code which fires the request and the backend.
Inside this interceptor, you can modify the request as you want to. As you can get the token via the lib, you do not need to pass the token into the interceptor.

> I need to tell my httpInterceptor file that the token is expired.
> I need to tell the httpInterceptor that the token is expired so that I can update the request headers for the endpoint.

So you do not have to tell the interceptor explicitly. If you want to implement the interceptor for yourself, then get the token from the lib via `getToken()` (V11) or `getAccessToken()` (V12) and set it into the authentication header. The token from the lib should be always the current token. 

Long story short: You do not pass the token to the interceptor or need to tell him, you let the interceptor read it from the lib.

Thanks

Fabian
